File: packages/snow_draw_engine/lib/draw/core/coordinates/coordinate_space.dart
import 'dart:math' as math;

import '../../types/draw_point.dart';

/// A coordinate space that can convert points to/from world coordinates.
///
/// This is primarily used for edit operations where we need to reason about a
/// rotated overlay (multi-select) or a rotated element.
abstract class CoordinateSpace {
  const CoordinateSpace();

  double get rotation;
  DrawPoint get origin;

  DrawPoint fromWorld(DrawPoint worldPoint);
  DrawPoint toWorld(DrawPoint localPoint);

  DrawPoint rotateVectorToWorld(DrawPoint localVector) {
    if (rotation == 0) {
      return localVector;
    }
    return _rotate(localVector.x, localVector.y, rotation);
  }

  DrawPoint rotateVectorToLocal(DrawPoint worldVector) {
    if (rotation == 0) {
      return worldVector;
    }
    return _rotate(worldVector.x, worldVector.y, -rotation);
  }

  /// Rotates [point] around [center] by [angle].
  DrawPoint rotatePoint({
    required DrawPoint point,
    required DrawPoint center,
    required double angle,
  }) {
    if (angle == 0) {
      return point;
    }
    final rotated = _rotate(point.x - center.x, point.y - center.y, angle);
    return DrawPoint(x: center.x + rotated.x, y: center.y + rotated.y);
  }

  DrawPoint _rotate(double x, double y, double angle) {
    final cosA = math.cos(angle);
    final sinA = math.sin(angle);
    return DrawPoint(x: x * cosA - y * sinA, y: x * sinA + y * cosA);
  }
}

File: packages/snow_draw_rust_engine/src/draw/core/coordinates/coordinate_space.rs
#![allow(dead_code)]

use crate::draw::types::draw_point::DrawPoint;

/// A coordinate space that can convert points to and from world coordinates.
///
/// This is primarily used for edit operations where we need to reason about a
/// rotated overlay (multi-select) or a rotated element.
pub trait CoordinateSpace {
    fn rotation(&self) -> f64;
    fn origin(&self) -> DrawPoint;

    fn from_world(&self, world_point: DrawPoint) -> DrawPoint;
    fn to_world(&self, local_point: DrawPoint) -> DrawPoint;

    fn rotate_vector_to_world(&self, local_vector: DrawPoint) -> DrawPoint {
        let rotation = self.rotation();
        if rotation == 0.0 {
            return local_vector;
        }

        rotate(local_vector.x, local_vector.y, rotation)
    }

    fn rotate_vector_to_local(&self, world_vector: DrawPoint) -> DrawPoint {
        let rotation = self.rotation();
        if rotation == 0.0 {
            return world_vector;
        }

        rotate(world_vector.x, world_vector.y, -rotation)
    }

    /// Rotates `point` around `center` by `angle`.
    fn rotate_point(&self, point: DrawPoint, center: DrawPoint, angle: f64) -> DrawPoint {
        if angle == 0.0 {
            return point;
        }

        let rotated = rotate(point.x - center.x, point.y - center.y, angle);
        DrawPoint::new(center.x + rotated.x, center.y + rotated.y)
    }
}

fn rotate(x: f64, y: f64, angle: f64) -> DrawPoint {
    let cos_a = angle.cos();
    let sin_a = angle.sin();
    DrawPoint::new(x * cos_a - y * sin_a, x * sin_a + y * cos_a)
}
